### ISSUE TITLE
LSP4J 0.11.0

### ISSUE THREAD

**Issue by jonahgraham** [eclipse/lsp4j#466]
This is the Release plan and TODO list for LSP4J release v0.11.0.

## Steps for Release

Items at the beginning of development

- [x] Create an Endgame Issue to track the release. As a starting point use [documentation/releasing.md](https://github.com/eclipse/lsp4j/blob/master/documentation/releasing.md).
    - [x] Add the [Endgame](https://github.com/eclipse/lsp4j/labels/endgame) label and Milestone for the release
    - [x] Make sure any previous edits made to [Endgame issues](https://github.com/eclipse/lsp4j/labels/endgame) of previous releases are updated in [documentation/releasing.md](https://github.com/eclipse/lsp4j/blob/master/documentation/releasing.md)
- [x] Ensure all previous [Endgame issues](https://github.com/eclipse/lsp4j/labels/endgame) are done.
- [x] Create a [New milestone](https://github.com/eclipse/lsp4j/milestones/new) for the release
- [x] Create release on [PMI](https://projects.eclipse.org/projects/technology.lsp4j)
- [x] Check [CHANGELOG.md](https://github.com/eclipse/lsp4j/blob/master/CHANGELOG.md) is up to date. The changelog should have a version entry, release date, API Breakages and other information consistent with current entries in the changelog.
- [x] Check [README.md](https://github.com/eclipse/lsp4j/blob/master/README.md) is up to date. In particular that the planned release and which versions of DAP and LSP are support is listed.
- [x] Increment version of all feature.xml, pom.xml and any other place full version is used. (Easiest way is global find and replace, e.g. `s/0.10.0/0.11.0/g` and review changes.) Ensure that `-SNAPSHOT` is restored in the [gradle/versions.gradle](https://github.com/eclipse/lsp4j/blob/master/gradle/versions.gradle) and  [releng/pom.xml](https://github.com/eclipse/lsp4j/blob/master/releng/pom.xml)
- [x] Enable `sh './releng/deploy-build.sh'` in [releng/build.Jenkinsfile](https://github.com/eclipse/lsp4j/blob/master/releng/build.Jenkinsfile) 
- [x] Ensure [the CI build](https://ci.eclipse.org/lsp4j/job/lsp4j-multi-build/job/master/) is stable - it is always better to release a "Green Dot" build

Items in the days ahead of Release day:

- [x] Schedule the release and if needed schedule a release review on the [PMI](https://projects.eclipse.org/projects/technology.lsp4j). A release review is needed every 12 months, not with each release.
- [x] Check [CHANGELOG.md](https://github.com/eclipse/lsp4j/blob/master/CHANGELOG.md) is up to date. The changelog should have a version entry, release date, API Breakages and other information consistent with current entries in the changelog.
- [x] Check [README.md](https://github.com/eclipse/lsp4j/blob/master/README.md) is up to date. In particular that the planned release and which versions of DAP and LSP are support is listed.
- [x] Check all closed PRs and Issues to make sure their milestone is set. (*Note:* this was not done before 0.11.0 release so many old PRs and Issues have no milestone, therefore only consider items back to approx 5 Nov 2020). [This search may be useful to identify such closed issues](https://github.com/eclipse/lsp4j/issues?q=is%3Aclosed+no%3Amilestone+updated%3A%3E%3D2020-11-06)

Items on Release day:

- [x] Prepare the repo for release by:
    - [x] removing `-SNAPSHOT` from [gradle/versions.gradle](https://github.com/eclipse/lsp4j/blob/master/gradle/versions.gradle)     
    - [x] removing `-SNAPSHOT` from [releng/pom.xml](https://github.com/eclipse/lsp4j/blob/master/releng/pom.xml) entries in `<dependencies>` section.
    - [x] disabling `sh './releng/deploy-build.sh'` in [releng/build.Jenkinsfile](https://github.com/eclipse/lsp4j/blob/master/releng/build.Jenkinsfile) 
    - see commit https://github.com/eclipse/lsp4j/commit/328ce8a4c89b0cd84fb62118f459b6cf79b09e90 for a past example
- [x] Push the above change
- [x] Run [the CI build](https://ci.eclipse.org/lsp4j/job/lsp4j-multi-build/job/master/)
- [x] Mark the build as Keep Forever and add to the description `v0.11.0`
- [x] Deploy the release by running [the Release CI job](https://ci.eclipse.org/lsp4j/job/lsp4j-release-eclipse) with parameters:
    - `LSP4J_PUBLISH_LOCATION` -> `updates/releases/0.11.0` ( <-- check version number)
    - `PROJECT` -> `lsp4j-multi-build/job/master`
    - `LSP4J_BUILD_NUMBER` -> the build that was just run above
- [x] Add to the deploy job description `v0.11.0`
- [x] Promote the staged repository to maven central
    - [Login to Nexus](https://oss.sonatype.org/#stagingRepositories)
        - To obtain permission add request to [OSSRH-26079](https://issues.sonatype.org/browse/OSSRH-26079)
    - go to *Staging Repositories*, after a short delay the staged LSP4J release should appear
    - click the staged LSP4J repo
    - press the *Close* button located in the toolbar. This runs activities, including checking various rules
    - once the rules are done (if successful), press the *Release* button
    - check https://search.maven.org/search?q=g:org.eclipse.lsp4j to make sure the latest release has arrived - this takes a while, 15 minutes for the files to be [on the server](https://repo1.maven.org/maven2/org/eclipse/lsp4j/org.eclipse.lsp4j/) and even longer for the [search indexes](https://search.maven.org/search?q=g:org.eclipse.lsp4j) to update
- [x] Update the meta-data on [PMI downloads page](https://projects.eclipse.org/projects/technology.lsp4j/downloads)
- [x] Tag the release. Example: `git tag -a v0.11.0 HEAD -m"LSP4J 0.11.0" && git push origin v0.11.0`
- [x] Contribute to Simrel. See [Simrel contribution example](https://git.eclipse.org/r/#/c/158624/)
- [x] Create a [release page on github](https://github.com/eclipse/lsp4j/releases/new)
- [x] Make an announcement on lsp4j-dev based on the [release page on github](https://github.com/eclipse/lsp4j/releases/tag/v0.11.0). [Example on lsp4j-dev archives](https://www.eclipse.org/lists/lsp4j-dev/msg00063.html)
- [x] Update [documentation/releasing.md](https://github.com/eclipse/lsp4j/blob/master/documentation/releasing.md) with any changes that may have been made to the release process.
- [x] Create the endgame for the next release right away, especially as version numbers and restoring `-SNAPSHOT` need to be done right away.

**Comment by jonahgraham** [comment]
This is the provisional endgame for 0.11.0. This release has not been scheduled yet, and even the version number may change.

**Comment by cdietrich** [comment]
we might also have to wait for guava update https://github.com/eclipse/lsp4j/issues/497

**Comment by cdietrich** [comment]
do we have a timeline for this? M3 is Feb 22th week and Downstream Users like Xtext and LSP4e may have to adapt until RC1 at March 1st week

**Comment by jonahgraham** [comment]
> do we have a timeline for this? M3 is Feb 22th week and Downstream Users like Xtext and LSP4e may have to adapt until RC1 at March 1st week

No - I did a review of open issues over the last 24 hours. I got no answer on [lsp4j-dev](https://www.eclipse.org/lists/lsp4j-dev/msg00085.html) and there are a number of items I marked as [0.11.0](https://github.com/eclipse/lsp4j/milestones/v0.11.0) because AFAIU they are issues with items in LSP 3.16 (Which 0.11.0 is supposed to support).

**Comment by cdietrich** [comment]
looks like mailing lists are broken. last mail i got is from Jab. 26th

**Comment by jonahgraham** [comment]
> looks like mailing lists are broken. last mail i got is from Jab. 26th

I am getting mail on other lists - AFAIK the last one sent on lsp4j-dev was the one I sent on Jan 26. I will ping on the list.

**Comment by cdietrich** [comment]
ahhh understood you did another mail recently

**Comment by cdietrich** [comment]
any update here? M3 is next week which would mean we should have a release/feature freeze then. or will we postpone to somewhen during 2021-06 simrel.

**Comment by jonahgraham** [comment]
There has been no response to my dev emails. I recommend just leaving 2020-03 simrel on 0.10.0.

**Comment by cdietrich** [comment]
ok. i am fine with that. i assume other clients like lsp4e would also struggle to update to 0.11.0 in a hurry.

**Comment by angelozerr** [comment]
For LemMinx (XML Language Server) we are very interested with 0.11.0 version which provides Linked Editing support. 

![image](https://user-images.githubusercontent.com/1932211/108513097-66de5200-72c2-11eb-8ceb-2e42eba10530.png)

For the moment we use SNAPSHOT version (https://github.com/eclipse/lemminx/pull/991) but we wait for LSP4J 0.11.0 release to merge the PR and create a release of XML Language Server.

Thank a lot for working on LSP4J 0.11.0 release!

**Comment by anpete** [comment]
Any update on this? 3.16 is done now and we are getting blocked on #478 

Thank you!

**Comment by jonahgraham** [comment]
Hi @anpete - we have the [following issues](https://github.com/eclipse/lsp4j/milestone/18) still assigned to 0.11.0, which IIUC are all related to finishing implementing 3.16.0 support in LSP4J. I will ping each of these issues to see if we can defer them to a 0.12.0 release.

**Comment by anpete** [comment]
@jonahgraham Thanks for the update.

**Comment by jonahgraham** [comment]
As of right now all 0.11.0 issues are marked as complete. I will schedule a release.

**Comment by jonahgraham** [comment]
[Release scheduled](https://projects.eclipse.org/projects/technology.lsp4j/releases/0.11.0) for this Friday and [intention announced on lsp4j-dev](https://www.eclipse.org/lists/lsp4j-dev/msg00088.html).

**Comment by jblievremont** [comment]
Hello, did you get around to releasing 0.11.0? I'm eagerly looking forward to being able to use the `Diagnostic#data` field :innocent:

**Comment by jonahgraham** [comment]
@jblievremont - release is imminent, unfortunately the time I had set aside for it on Friday was when GitHub had a major outage.  https://www.eclipse.org/lists/lsp4j-dev/msg00089.html - sorry for the delay I am trying to find time to reschedule as soon as possible, but first I have to handle the 2021-03 Eclipse IDE simultaneous release which should conclude tomorrow.

If another committer wants to walk through the checklist in the first comment on this thread that would be fine by me :-)

**Comment by jonahgraham** [comment]
[Release re-scheduled](https://projects.eclipse.org/projects/technology.lsp4j/releases/0.11.0) for this Friday (the 19th) and [intention announced on lsp4j-dev](https://www.eclipse.org/lists/lsp4j-dev/msg00090.html).

**Comment by jonahgraham** [comment]
https://github.com/eclipse/lsp4j/releases/tag/v0.11.0 just went live - but I am actually finishing up some small steps (like simrel) so some updates to it are coming soon.

**Comment by jonahgraham** [comment]
The [release](https://github.com/eclipse/lsp4j/releases/tag/v0.11.0) is now done.

The [maven search index](https://search.maven.org/search?q=g:org.eclipse.lsp4j) has not refreshed yet, but the files are on the [maven central server](https://repo1.maven.org/maven2/org/eclipse/lsp4j/org.eclipse.lsp4j/).
